File: day05/js/solution.js
import fs from "fs";

const input = fs.readFileSync("../input.txt", "utf-8").trimEnd();
const [rangesBlock, idsBlock] = input.split(/\r?\n\s*\r?\n/);

const ranges = rangesBlock
  .split("\n")
  .map(line => line.split("-").map(Number));

const ids = idsBlock
  .split("\n")
  .map(Number);

function isFresh(id, ranges) {
  return ranges.some(([start, end]) => id >= start && id <= end);
}

function part1(ranges, ids) {
  return ids.filter(id => isFresh(id, ranges)).length;
}

function part2(ranges) {
  ranges.sort((a, b) => a[0] - b[0]);

  let total = 0;
  let [curStart, curEnd] = ranges[0];

  for (let i = 1; i < ranges.length; i++) {
    const [start, end] = ranges[i];

    if (start <= curEnd + 1) {
      curEnd = Math.max(curEnd, end);
    } else {
      total += curEnd - curStart + 1;
      [curStart, curEnd] = [start, end];
    }
  }

  total += curEnd - curStart + 1;

  return total;
}

console.log("Part 1:", part1(ranges, ids));
console.log("Part 2:", part2(ranges));

File: day05/py/solution.py
from pathlib import Path
import re

text = Path("../input.txt").read_text().strip()
ranges_block, ids_block = re.split(r"\r?\n\s*\r?\n", text)

ranges = [
    tuple(map(int, line.split("-")))
    for line in ranges_block.splitlines()
]

ids = list(map(int, ids_block.splitlines()))


def is_fresh(id_, ranges):
    return any(start <= id_ <= end for start, end in ranges)


def part1(ranges, ids):
    return sum(1 for id_ in ids if is_fresh(id_, ranges))


def part2(ranges):
    ranges.sort()
    total = 0

    cur_start, cur_end = ranges[0]

    for start, end in ranges[1:]:
        if start <= cur_end + 1:
            cur_end = max(cur_end, end)
        else:
            total += cur_end - cur_start + 1
            cur_start, cur_end = start, end

    total += cur_end - cur_start + 1
    return total

print("Part 1:", part1(ranges, ids))
print("Part 2:", part2(ranges))
